### PR TITLE
feat: add new `__list__` parameter for list-based array types

### PR DESCRIPTION
**Pull request by agoose77**
This PR adds a new `__list__` nominal type parameter to associate with behaviors. It should only be added to lists. In addition to setting an array-class for `ListOffsetArray` (et al.) via `layout.parameter("__list__")`, we can also lookup the list in a purelist sense, allowing list-of-type to have an array class too.

Setting `__list__` and `__record__` an error for non-intended types, but that should be another PR (there are fixes that need to be made to both `validity_error` and broadcasting).

### CODE REVIEW

**Comment by agoose77** [comment]
The addition of the new `__list__` nominal type parameter provides the ability to set `__typestr__` for a list type. However, unlike records which can sensibly be considered atoms, lists are usually part of a deeper structure that we care about seeing the type of. Having
`N * my-list` instead of `N * my-list * int64` is worse. I wonder if we should think about this before merging this PR.

File: docs/reference/ak.behavior.md
Comment (jpivarski) [inline docs/reference/ak.behavior.md]: This can also be implemented as

```python
def reversed(self):
    return self[..., ::-1]
```

but the original is fine: it's some computation that one might want to do—it doesn't matter what that computation is.

File: src/awkward/_parameters.py
Comment (jpivarski) [inline src/awkward/_parameters.py:7]: :heavy_check_mark:

**Review by jpivarski — APPROVED** [review]
Very nice! I like the new documentation examples, too.

Since this PR is more complicated and you might think of something you need to add at the last minute, I'll let you merge it. I think it's done.

File: docs/reference/ak.behavior.md
Comment (agoose77) [reply docs/reference/ak.behavior.md]: Indeed, I hesitated on adding this. Originally I was going to add a string-related method that inverted the character order. I'm going to change this because it is not an important part of the messaging.

Comment (agoose77) [reply docs/reference/ak.behavior.md:525]:
```suggestion
    def reversed(self):
        return self[..., ::-1]
```

Comment (agoose77) [reply docs/reference/ak.behavior.md]:
```suggestion
    def reversed(self):
        return self[..., ::-1]
```

**Comment by jpivarski** [comment]
> The addition of the new `__list__` nominal type parameter provides the ability to set `__typestr__` for a list type. However, unlike records which can sensibly be considered atoms, lists are usually part of a deeper structure that we care about seeing the type of. Having `N * my-list` instead of `N * my-list * int64` is worse. I wonder if we should think about this before merging this PR.

Sorry that I missed this question!

A `__typestr__` for `__list__` would be useful because a common reason to override a list would be to define an opaque binary blob. The developer of that library would want the type of the hidden data to show up as "`my-list`", _without_ any hint about the contents. So, for instance,

```
1000 * var * PNG-image
```

for an array of lists of overloaded bytestrings that are interpreted as PNG images.

The `__typestr__` was introduced to make it possible to label strings as `string` and adhere to Datashape, as part of the goal of making string entirely user-definable. I think that was already gone at the time when Awkward 1 was converted to Awkward 2.

Since then, the `__typestr__` was useful for overloaded records, in which the library developer wanted to hide the details of the record structure. (I think this was in some version of Coffea.)

The bytestring use-case (I can't think of any others, involving `__list__`) would also want to hide the internal structure of the list, for the same reason. In fact, that was also true when `__typestr__` was set to `"string"`.

So yes, I think it would be a good addition, but it can be applied in the same way that it is for `__record__`, which is simpler than trying to let the inner contents through (or parameterize/templatize `__typestr__` in some way).

**Pull request merged**
